feat(i18n): issue #27

salvatorericcardi opened the pull request:
# 📚 Description
## What
<!-- Explain what this PR addresses: what this will do once merged? -->
You should be able to switch between english and italian language pathname. Based on lang path the code detect the active language and fix the url. Added redirect from old page without localization to new localized url.

## Why
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
I implemented the feature following the add internazionalization issue #27 

# 🧪 Testing Guide
<!-- A testing plan to verify this code is very appreciated. Add here the steps to follow to verify the code -->
Run `npm run dev` and switch between languages by clicking on the swticher (not the flags). Follow the admin and feedback links to watch the url localized as well.

# 🖼 Screenshots/Recordings
<!-- Do you have any images/videos to attach to this PR as asset in order to add more context for reviewer? -->
![image](https://github.com/latina-in-tech/latina-in-tech.github.io/assets/29570607/59a768ad-858c-4922-867d-8bf82a836697)

# 📝 To Do
<!-- Do you have left some known TODOs in the code, that you can not solve now? IF taht's the case, please create an issue for those and reference them here-->
The LocaleSwitcher component has an issue after few clicks it render twice causing problems with language switch (look the comments under LocaleSwitcher.tsx:29). To do: avoiding the re-rendering of the component on router event "routechangecomplete". Look at [nextjs router events page](https://nextjs.org/docs/pages/api-reference/functions/use-router#routerevents )

Undermaken reviewed (changes requested):
Here some minor UI adjustements

![image](https://github.com/latina-in-tech/latina-in-tech.github.io/assets/822471/7b9bedfb-1297-4264-bb6c-ed2b8872160a)

coluzziandrea commented on src/pages/admins/team.tsx:
```diff
@@ -138,29 +139,11 @@ const AdminCard: React.FC<Admin> = ({
-    </div>
-  );
+  useEffect(() => {
+    window.location.href = `/${i18n.defaultLocale}`;
```
Actually it was me to suggest him to leave the old routes and re-direct from them to the new ones 😅 Do you think it is fine to break the URLs? ALso another thing I can think of, when you type an exact URL on a big site, you will not get a 404, but you will be just redirected to the default language.

Perhaps we can use a query parameter instead of the path itself? i.e. `/admins/team?language=en` instead of `/en/admins/team`

Undermaken replied:
ok, it makes sense. I was not considering the event pages that are been spread in several sites/posts
I'm wondering if we can have a clearer solution in order to avoid components that render nothing (it's a kind of anti pattern) and code duplication (I'm figuring a new contributor that meets a duplicated resources for each resource in pages directory)

What about to use a backward compatibility hook in the `App` component?
In that way we do one thing in one place and we can remove it when we'll be pretty sure users can't access old ulrs

something like this

```ts
const oldPaths = [
  '/admins/team',
  '/events/\\d+',
  '/feedback/new',
  '/newsletter',
  '/community'
].map(p => new RegExp(`^${p}`));
const useBackRouteCompatibility = () => {
  const router = useRouter();
  React.useEffect(() => {
    if (oldPaths.some(p => p.test(router.asPath))) {
      void router.replace(
        router.asPath,
        `/${i18n.defaultLocale}${router.asPath}`
      );
    }
  }, [router]);
};

[...]

export default function App({ Component, pageProps }: AppProps) {
  useBackRouteCompatibility();
  return (
      <Component {...pageProps} />
  );
}

```
---

btw, no strictly related this, with this implementation if a user navigates LiT sites and click on an event, he will be redirected always to the home page

https://github.com/latina-in-tech/latina-in-tech.github.io/assets/822471/1c288bc5-a3f7-40c2-b4bd-4cb3a3689664


---

salvatorericcardi commented:
Here a list of the last fixations:
- ui adjustments
- single event links to the right page
- router.replace replaced router.push

I've not implemented backroutecompatibility func because we will have a 404 error by removing the first index.tsx in the pages dir but I inserted the code with comments in _app.tsx so you can evaluate the situation.

Undermaken reviewed (changes requested):
@salvatorericcardi thank you again for your precious contribution ❤️ 
I guess we are almost there! this issue it's not so easy!

Maybe I found a solution about redirect.
Instead of components that redirects or an hook that does the same, nextjs provide a redirect configuration, see here https://nextjs.org/docs/pages/api-reference/next-config-js/redirects

So I tested these lines into `next.config.js` and it seems to work properly
```ts
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: 'export',
  reactStrictMode: true,
  images: {
    unoptimized: true
  },
  async redirects() {
    return [
      '/',
      '/admins/team',
      '/events/:eventId',
      '/feedback/new',
      '/newsletter',
      '/community'
    ].map(r => ({
      source: r,
      destination: '/it' + r,
      permanent: true
    }));
  }
};

module.exports = nextConfig;
```

So if you find this solution valueable, here a little recap about files that can be deleted:
we should keep only these file in pages directory, all other can be removed

![Screenshot 2024-04-16 alle 21 38 06](https://github.com/latina-in-tech/latina-in-tech.github.io/assets/822471/27963c9a-9d57-487d-8fb0-29d923618a98)

---

here another little bug I found while browsing the pages

https://github.com/latina-in-tech/latina-in-tech.github.io/assets/822471/4ddd9010-2482-4f96-9525-2ef3c786a6c7

---

at the end here another point about navigation: if I start browsing the site with `it` and then change to 'en', my history is made by a mix of them, here an example

https://github.com/latina-in-tech/latina-in-tech.github.io/assets/822471/67b7c0dc-5def-4860-bc6e-9afe5bbc4457

I'm not really sure it's a bug. Maybe your are more expert than me. I leave you any consideration

salvatorericcardi commented:
@Undermaken I guess the first bug is caused by this code, precisely in the href attribute on the link component.

```
{links.map(item => (
  <Link
    key={item.name}
    href={
      // if condition is true print item.href otherwise is admin team link and fix the url
      item.name !== 'Admin team' 
        ? item.href
        : fix(item.href)
    }
    target={item.local ? undefined : '_blank'}
    className={classNames(
      item.current
        ? 'bg-slate-700 text-white dark:bg-black dark:bg-opacity-4 dark:text-slate-300'
        : 'text-slate-600 hover:bg-slate-200 dark:text-slate-300 dark:hover:bg-black dark:hover:bg-opacity-20',
      'flex items-center gap-2 rounded-md px-2 py-2 text-sm font-medium'
    )}
    aria-current={item.current ? 'page' : undefined}
  >
    <item.icon fontSize={24} />
  </Link>
  ))}
```

Instead, I can't understand the second bug. It seems to render the page twice but can you try to explain me the bug in words? I'll go deeper asap.

salvatorericcardi commented:
Warning: [https://nextjs.org/docs/messages/export-no-custom-routes](https://nextjs.org/docs/messages/export-no-custom-routes)

**Redirects don't work automatically with static export only in local environment.** 
This is why I remember I pushed with duplicated pages and windows.location.href (manual redirects), initially.

Undermaken commented:
> Warning: https://nextjs.org/docs/messages/export-no-custom-routes
> 
> Verify if the link content is valid even for pages router or only for app router. I had problems to deploy my website with nextjs app router.
> 
> **Redirects don't work automatically with static export only in local environment.** This is why I remember I pushed with duplicated pages and windows.location.href (manual redirects), initially.

@salvatorericcardi I apologise for the delay in this review. These are very busy days at work.
btw you spotted the reason why we can't adopt the solution I suggested before, thank you for going deeper!

I just rolled back on the usage of `useBackRouteCompatibility` hooks.
Within your last changes it seems to work properly, can you confirm? If also for you is ok, I guess we may merge this PR

---

> at the end here another point about navigation: if I start browsing the site with it and then change to en, my history is made by a mix of them, here an example

about this bug, I'm not really sure it is a bug. I try to explain better, but I guess we can follow up it in a dedicated PR.
When the user navigates the web site, maybe at some point, he could change the language, so let's suppose this scenario

- users opens the home page 🇮🇹 
- go the eventX 🇮🇹 
- go to the admin team 🇮🇹 
- change the language 🇬🇧 
- go to the eventY 🇬🇧 
- go the the admin team 🇬🇧 
- go back in the browser history
- go to the eventY 🇬🇧 
- go back in the browser history
- go to the admin team 🇮🇹 <---- 🐛 

The user loads a page in 🇮🇹 even if he hasn't explicitly changed the language.

But as I said before, not sure if this is really a bug, in any case leave it for future throughts.

salvatorericcardi commented:
@Undermaken you're right, it is not a bug. This happens because when we visit a page the browser write this page in its history so we can access it by going back. Btw this can be improved with a dedicated pr. Now keep focused on the i18n mvp version.

Undermaken commented:
> @Undermaken you're right, it is not a bug. This happens because when we visit a page the browser write this page in its history so we can access it by going back. Btw this can be improved with a dedicated pr. Now keep focused on the i18n mvp version.
> 
> PS Don't worry take your time to respond. I will test code asap today so you can merge and close the pr.

Great!
When also your tests went good, let me know!

Two last minor question
1. Can I celebrate your contribution in the Telegram channel? This may be important to increase the community traction for this project.
2. If you like, it could be valuable to have a PR where you add even one translation. This might be useful for other contributors to have a starting point.

coluzziandrea commented on src/components/LocaleSwitcher.tsx:
```diff
@@ -0,0 +1,65 @@
+      const regex = /it|en/;
+
+      if (regex.test(pathname)) {
+        // not works: onclick runs twice ######
```
Hi @salvatorericcardi! If for some reason you're going to change the code again, I'd suggest you to remove useless 'console.log' comments here and there.

coluzziandrea commented on src/pages/_document.tsx:
```diff
@@ -5,7 +5,7 @@ export default function Document() {
   return (
     <Html>
       <Head />
-      <body className='dark:bg-slate-900 dark:text-slate-300' lang='it'>
```
can we extract the lang from the path here?

coluzziandrea commented:
great work so far @salvatorericcardi !!!🔥

salvatorericcardi commented:
@Undermaken I moved redirects like your suggestion, but they seem not working. Maybe it is my local error or a stupid mistake. It's weird because for a little moment automatic redirects seemed working, not now. 
1. Of course, you can celebrate my contribution from its merge with main branch. First of all, we need to complete the pr.

@coluzziandrea I can't move lang attribute in _document, but it is a minor fixation I think, or not? Even, I removed useless comments.

For you all, try to test because if the code works we can merge then think to translations and follow ups otherwise the problem with redirects is a global issue I will restore redirects page by page but I don't think it is necessary.

coluzziandrea commented:
@salvatorericcardi sorry, I was working on it too ahahaha. I've added you as a community member and added some translations token. For me it is ok to be merged! 

Great work and many thanks!!!
